File: app/src/main/java/com/futsch1/medtimer/medicine/MedicinesFragment.java
package com.futsch1.medtimer.medicine;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;

import android.app.AlertDialog;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.text.Editable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.futsch1.medtimer.MedicineViewModel;
import com.futsch1.medtimer.OptionsMenu;
import com.futsch1.medtimer.R;
import com.futsch1.medtimer.database.Medicine;
import com.futsch1.medtimer.helpers.DeleteHelper;
import com.futsch1.medtimer.helpers.SimpleIdlingResource;
import com.futsch1.medtimer.helpers.SwipeHelper;
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

public class MedicinesFragment extends Fragment {
    private SimpleIdlingResource idlingResource;
    @SuppressWarnings("java:S1450")
    private MedicineViewModel medicineViewModel;
    @SuppressWarnings("java:S1450")
    private MedicineViewAdapter adapter;
    private HandlerThread thread;
    private OptionsMenu optionsMenu = null;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.thread = new HandlerThread("GetSummary");
        this.thread.start();
        idlingResource = new SimpleIdlingResource(MedicinesFragment.class.getName());
        idlingResource.setBusy();

        // Get a new or existing ViewModel from the ViewModelProvider.
        medicineViewModel = new ViewModelProvider(this).get(MedicineViewModel.class);

        optionsMenu = new OptionsMenu(this,
                medicineViewModel,
                NavHostFragment.findNavController(this), false);

        adapter = new MedicineViewAdapter(thread, requireActivity(), medicineViewModel.medicineRepository);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View fragmentView = inflater.inflate(R.layout.fragment_medicines, container, false);
        // Medicine recycler
        RecyclerView recyclerView = fragmentView.findViewById(R.id.medicineList);

        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(fragmentView.getContext()));

        // Swipe to delete
        ItemTouchHelper itemTouchHelper = SwipeHelper.createSwipeHelper(requireContext(), viewHolder -> deleteItem(requireContext(), viewHolder.getItemId(), viewHolder.getBindingAdapterPosition()), adapter);
        itemTouchHelper.attachToRecyclerView(recyclerView);

        postponeEnterTransition();

        setupAddMedicineButton(fragmentView);

        MedicinesMenu medicinesMenu = new MedicinesMenu(medicineViewModel, thread);
        requireActivity().addMenuProvider(medicinesMenu, getViewLifecycleOwner());

        // Connect view model to recycler view adapter
        medicineViewModel.getMedicines().observe(getViewLifecycleOwner(), l -> {
            adapter.submitList(l);
            medicinesMenu.medicinesList = l;
            startPostponedEnterTransition();
            idlingResource.setIdle();
        });

        requireActivity().addMenuProvider(optionsMenu, getViewLifecycleOwner());

        return fragmentView;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (optionsMenu != null) {
            optionsMenu.onDestroy();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (thread != null) {
            thread.quit();
        }
        if (idlingResource != null) {
            idlingResource.destroy();
        }
    }

    private void deleteItem(Context context, long itemId, int adapterPosition) {
        DeleteHelper deleteHelper = new DeleteHelper(context);
        deleteHelper.deleteItem(R.string.are_you_sure_delete_medicine, () -> medicineViewModel.medicineRepository.deleteMedicine((int) itemId), () -> adapter.notifyItemChanged(adapterPosition));
    }

    private void setupAddMedicineButton(View fragmentView) {
        ExtendedFloatingActionButton fab = fragmentView.findViewById(R.id.addMedicine);
        fab.setOnClickListener(view -> {
            TextInputLayout textInputLayout = new TextInputLayout(requireContext());
            TextInputEditText editText = new TextInputEditText(requireContext());
            editText.setLayoutParams(new LinearLayout.LayoutParams(MATCH_PARENT, WRAP_CONTENT));
            editText.setHint(R.string.medicine_name);
            editText.setSingleLine();
            editText.setId(R.id.medicineName);
            textInputLayout.addView(editText);

            AlertDialog.Builder builder = getAlertBuilder(textInputLayout, editText);
            AlertDialog dialog = builder.create();
            dialog.show();
        });
    }

    @NonNull
    private AlertDialog.Builder getAlertBuilder(TextInputLayout textInputLayout, TextInputEditText editText) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setView(textInputLayout);
        builder.setTitle(R.string.add_medicine);
        builder.setPositiveButton(R.string.ok, (dialog, which) -> {
            Editable e = editText.getText();
            if (e != null) {
                new Handler(thread.getLooper()).post(() -> {
                    double highestSortOrder = medicineViewModel.medicineRepository.getHighestMedicineSortOrder();
                    Medicine medicine = new Medicine(e.toString().trim());
                    medicine.sortOrder = highestSortOrder + 1;
                    int medicineId = (int) medicineViewModel.medicineRepository.insertMedicine(medicine);
                    requireActivity().runOnUiThread(() -> navigateToMedicineId(medicineId));
                });
            }
        });
        builder.setNegativeButton(R.string.cancel, (dialog, which) -> dialog.dismiss());
        return builder;
    }

    private void navigateToMedicineId(int medicineId) {
        NavController navController = Navigation.findNavController(this.requireView());
        MedicinesFragmentDirections.ActionMedicinesFragmentToEditMedicineFragment action = MedicinesFragmentDirections.actionMedicinesFragmentToEditMedicineFragment(
                medicineId
        );
        navController.navigate(action);
    }
}

File: app/src/main/java/com/futsch1/medtimer/medicine/MedicinesFragment.kt
package com.futsch1.medtimer.medicine

import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.core.view.setPadding
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.Navigation.findNavController
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.futsch1.medtimer.MedicineViewModel
import com.futsch1.medtimer.OptionsMenu
import com.futsch1.medtimer.R
import com.futsch1.medtimer.database.FullMedicine
import com.futsch1.medtimer.database.Medicine
import com.futsch1.medtimer.database.MedicineRepository
import com.futsch1.medtimer.di.Dispatcher
import com.futsch1.medtimer.di.MedTimerDispatchers
import com.futsch1.medtimer.helpers.DeleteHelper
import com.futsch1.medtimer.helpers.SimpleIdlingResource
import com.futsch1.medtimer.helpers.SwipeHelper
import com.futsch1.medtimer.helpers.dpToPx
import com.futsch1.medtimer.helpers.showSoftKeyboard
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

@AndroidEntryPoint
class MedicinesFragment : Fragment() {
    @Inject
    @Dispatcher(MedTimerDispatchers.IO)
    lateinit var dispatcher: CoroutineDispatcher

    @Inject
    @Dispatcher(MedTimerDispatchers.Main)
    lateinit var mainDispatcher: CoroutineDispatcher

    @Inject
    lateinit var medicineRepository: MedicineRepository

    @Inject
    lateinit var medicinesMenu: MedicinesMenu

    private lateinit var idlingResource: SimpleIdlingResource
    private val medicineViewModel: MedicineViewModel by viewModels()
    private lateinit var adapter: MedicineViewAdapter

    @Inject
    lateinit var optionsMenuFactory: OptionsMenu.Factory

    @Inject
    lateinit var medicineViewAdapterFactory: MedicineViewAdapter.Factory
    private lateinit var optionsMenu: OptionsMenu

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        idlingResource = SimpleIdlingResource(MedicinesFragment::class.java.getName())
        idlingResource.setBusy()

        optionsMenu = optionsMenuFactory.create(
            this,
            NavHostFragment.findNavController(this),
            false,
            medicineViewModel
        )

        adapter = medicineViewAdapterFactory.create(requireActivity())
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val fragmentView = inflater.inflate(R.layout.fragment_medicines, container, false)
        // Medicine recycler
        val recyclerView = fragmentView.findViewById<RecyclerView>(R.id.medicineList)

        recyclerView.setAdapter(adapter)
        recyclerView.setLayoutManager(LinearLayoutManager(fragmentView.context))

        // Swipe to delete
        val itemTouchHelper = SwipeHelper.createSwipeHelper(
            requireContext(),
            { viewHolder ->
                deleteItem(
                    requireContext(),
                    viewHolder.itemId,
                    viewHolder.getBindingAdapterPosition()
                )
            },
            adapter
        )
        itemTouchHelper.attachToRecyclerView(recyclerView)

        postponeEnterTransition()

        setupAddMedicineButton(fragmentView)

        requireActivity().addMenuProvider(medicinesMenu, getViewLifecycleOwner())

        // Connect view model to recycler view adapter
        viewLifecycleOwner.lifecycleScope.launch {
            medicineViewModel.medicines.collect { l: List<FullMedicine> ->
                adapter.submitList(l)
                medicinesMenu.medicines = l
                startPostponedEnterTransition()
                idlingResource.setIdle()
            }
        }

        requireActivity().addMenuProvider(optionsMenu, getViewLifecycleOwner())

        return fragmentView
    }

    override fun onDestroy() {
        super.onDestroy()
        idlingResource.destroy()
        optionsMenu.onDestroy()
    }

    private fun deleteItem(context: Context, itemId: Long, adapterPosition: Int) {
        DeleteHelper.deleteItem(
            context,
            R.string.are_you_sure_delete_medicine,
            { lifecycleScope.launch { medicineRepository.delete(itemId.toInt()) } },
            { adapter.notifyItemChanged(adapterPosition) })
    }

    private fun setupAddMedicineButton(fragmentView: View) {
        val fab = fragmentView.findViewById<ExtendedFloatingActionButton>(R.id.addMedicine)
        fab.setOnClickListener { _: View? ->
            val textInputLayout = TextInputLayout(requireContext())
            val editText = TextInputEditText(requireContext())
            editText.setLayoutParams(LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))
            editText.setHint(R.string.medicine_name)
            editText.setSingleLine()
            editText.setId(R.id.medicineName)
            editText.postDelayed({ editText.showSoftKeyboard() }, 200)

            textInputLayout.addView(editText)
            textInputLayout.setPadding(requireContext().resources.dpToPx(16f).toInt())

            val builder = getAlertBuilder(textInputLayout, editText)
            val dialog = builder.create()
            dialog.show()
        }
    }

    private fun getAlertBuilder(textInputLayout: TextInputLayout?, editText: TextInputEditText): MaterialAlertDialogBuilder {
        return MaterialAlertDialogBuilder(requireContext())
            .setView(textInputLayout)
            .setTitle(R.string.add_medicine)
            .setPositiveButton(R.string.ok) { _: DialogInterface?, _: Int ->
                val text = editText.getText() ?: return@setPositiveButton

                lifecycleScope.launch(dispatcher) {
                    val highestSortOrder = medicineRepository.getHighestSortOrder()
                    val medicine = Medicine(text.toString().trim())
                    medicine.sortOrder = highestSortOrder + 1
                    val medicineId = medicineRepository.create(medicine).toInt()
                    withContext(mainDispatcher) { navigateToMedicineId(medicineId) }
                }
            }
            .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
    }

    private fun navigateToMedicineId(medicineId: Int) {
        val navController = findNavController(this.requireView())
        val action = MedicinesFragmentDirections.actionMedicinesFragmentToEditMedicineFragment(
            medicineId
        )
        navController.navigate(action)
    }
}